contracts/Aavegotchi/facets/WhitelistFacet.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.1;
+
+import {LibMeta} from "../../shared/libraries/LibMeta.sol";
+import {Modifiers, Whitelist} from "../libraries/LibAppStorage.sol";
+
+contract WhitelistFacet is Modifiers {
+    event WhitelistCreated(uint256 indexed whitelistId);
+    event WhitelistUpdated(uint256 indexed whitelistId);
+
+    function createWhitelist(string calldata _name, address[] calldata _whitelistAddresses) external {
+        uint256 whitelistLength = _whitelistAddresses.length;
+        require(whitelistLength > 0, "WhitelistFacet: Whitelist length should be larger than zero");
+
+        address sender = LibMeta.msgSender();
+        uint256 whitelistId = s.whitelists.length + 1;
+        address[] memory addresses;
+        Whitelist memory whitelist = Whitelist({owner: sender, name: _name, addresses: addresses});
+        s.whitelists.push(whitelist);
+        _addAddressesToWhitelist(whitelistId, _whitelistAddresses);
+
+        emit WhitelistCreated(whitelistId);
+    }
+
+    function updateWhitelist(uint256 _whitelistId, address[] calldata _whitelistAddresses) external {
+        address sender = LibMeta.msgSender();
+
+        require((s.whitelists.length >= _whitelistId) && (_whitelistId > 0), "WhitelistFacet: whitelist not found");
+        require(s.whitelists[_whitelistId - 1].owner == sender, "WhitelistFacet: not whitelist owner");
+
+        uint256 whitelistLength = _whitelistAddresses.length;
+        require(whitelistLength > 0, "WhitelistFacet: Whitelist length should be larger than zero");
+
+        _addAddressesToWhitelist(_whitelistId, _whitelistAddresses);
+
+        emit WhitelistUpdated(_whitelistId);
+    }
+
+    function getWhitelist(uint256 _whitelistId) external view returns (Whitelist memory) {
+        require((s.whitelists.length >= _whitelistId) && (_whitelistId > 0), "WhitelistFacet: whitelist not found");
+        return s.whitelists[_whitelistId - 1];
+    }
+
+    function getWhitelists() external view returns (Whitelist[] memory) {
+        return s.whitelists;
+    }
+
+    function _addAddressesToWhitelist(uint256 _whitelistId, address[] calldata _whitelistAddresses) internal {
+        for (uint256 i; i < _whitelistAddresses.length; i++) {
+            if (!s.isWhitelisted[_whitelistId][_whitelistAddresses[i]]) {
+                s.whitelists[_whitelistId - 1].addresses.push(_whitelistAddresses[i]);
+                s.isWhitelisted[_whitelistId][_whitelistAddresses[i]] = true;
```

Comment:
An event with this info would probably be useful

Reply:
It might be thousands of addresses, so would be very burdensome. For maintaining the subgraph, you would update the addresses by calling getWhitelist()

Reply:
Adding a 1000 or more addresses would cost more than 40 million gas which would exceed the Polygon block gas limit (30 million), so addresses would need to be added in multiple transactions.   One possibility if it is needed in the future is to add an upgrade that uses a merkle tree for validating people.

Reply:
Yeah I think a merkle tree would be a good option, if Guilds require thousands of users. Will keep that in mind for a future upgrade.